src/genai/eyfs/run_pinecone_index.py

```diff
@@ -0,0 +1,130 @@
+"""Join the BBC activities with the labelled activities and build a Chroma index."""
+
+import os
+
+from typing import Generator
+
+import numpy as np
+import openai
+import pandas as pd
+import pinecone
+
+from dotenv import load_dotenv
+
+from genai.eyfs import get_embedding
+from genai.utils import batch
+from genai.utils import read_json
+
+
+load_dotenv()
+
+
+openai.api_key = os.environ["OPENAI_API_KEY"]
+PATH_TO_LABELLED_ACTIVITIES = "data/eyfs_labels/parsed_json.jsonl"
+PATH_TO_BBC_ACTIVITIES = "data/eyfs/tiny_happy_people - final - tiny_happy_people - final.csv"
+INDEX_NAME = "eyfs-index"
+
+
+def get_labelled_bbc_activities(path: str) -> pd.DataFrame:
+    """Read and clean the labelled BBC activities file and return a dataframe."""
+    data = read_json(path, lines=True)
+    df = pd.concat([pd.DataFrame([line]) for line in data])
+    df["prediction"] = df["prediction"].apply(lambda row: row if row else np.nan)
+    df = df[~df.prediction.isnull()]
+    df = df[df.prediction != "None"]
+    df = df.rename({"prediction": "areas_of_learning", "url": "URL"}, axis=1)
+
+    return df
+
+
+def get_bbc_activities(path: str) -> pd.DataFrame:
+    """Read and clean the BBC activities file and return a dataframe."""
+    df = pd.read_csv(path)
+    df = df.rename(columns={"Age Range (if applicable)": "Age", "CONTENT": "title"})
+    df = df.dropna(subset=["text", "URL"])
+    df = df.drop_duplicates(subset=["URL"])
+
+    return df
+
+
+def build_vector_index(
+    items: Generator,
+    index_name: str,
+    vector_size: int = 1536,
+    metric: str = "euclidean",
+    environment: str = "us-west1-gcp",
+) -> None:
+    """Build a vector index from a dataframe.
+
+    Parameters
+    ----------
+    items
+        The items to be indexed.
+
+    index_name
+        Name of the index.
+
+    vector_size
+        Length of the indexed vectors.
+
+    metric
+        The distance metric to use.
+
+    environment
+        The cloud environment to use.
+
+    """
+    # Connect to pinecone
+    pinecone.init(
+        api_key=os.environ["PINECONE_API_KEY"],
+        environment=environment,
+    )
+
+    # Search by this metadata
+    metadata_config = {"indexed": ["areas_of_learning"]}
+
+    # Build a new index every time
+    if index_name in pinecone.list_indexes():
+        pinecone.delete_index(index_name)
```

Comment:
When do you create the index generally? 

Is this every time when restarting the app, or when you update the data... or alternatively can we create it once and can you just keep adding to the index?

Reply:
Currently, we clean and encode the BBC data, build the index and index documents once and offline (ie locally, before launching the app). Then we just connect to pinecone and call the index. 

This has a few shortcomings:
- manual effort (we need to run the `run_pinecone_index.py` script to build the index)
- you cannot just insert new documents, you must create the index from scratch

Ideally, those tasks would be separated completely:
- Clean the BBC data
- Encode the texts
- Build the index if it does not exist and add documents, otherwise upsert any new items

I was planning on doing that in a follow-up PR (I wanted us to settle on a vector db provider first). I will tag Solomon to the next PR too in case he wants to use it in his work

Reply:
Learned a new word "upsert" today, awesome